### ISSUE TITLE
[BUG] Dialog module start up

### ISSUE THREAD

**Issue by volga629-1** [OpenSIPS/opensips#2567]
Dialog module with PgSQL 12.6 is preventing opensips from starting properly.
I tested with mode 1 and 2 both are failing, only 0 starts OK.

Version

```
[root@dev1-msg opensips]# opensips -V
version: opensips 3.2.0-beta (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, USE_MCAST, SHM_MMAP, PKG_MALLOC, Q_MALLOC, F_MALLOC, HP_MALLOC, DBG_MALLOC, CC_O0, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
main.c compiled on 00:00:00 Jun  3 2021 with gcc 10
```

Error

```
Jul 06 20:55:49 dev1-msg.networklab.support /usr/sbin/opensips[969]: CRITICAL:dialog:dlg_connect_db: BUG - db connection found already open
Jul 06 20:55:49 dev1-msg.networklab.support /usr/sbin/opensips[969]: ERROR:dialog:child_init: failed to connect to database (rank=-2)
Jul 06 20:55:49 dev1-msg.networklab.support /usr/sbin/opensips[969]: ERROR:core:init_mod_child: failed to initializing module dialog, rank -2
Jul 06 20:55:49 dev1-msg.networklab.support /usr/sbin/opensips[969]: ERROR:core:start_module_procs: error in init_child for PROC_MODULE
```

Config

```
#### Dialog
loadmodule "dialog.so"
#modparam("dialog", "db_url", "postgres://")
modparam("dialog", "db_mode", 1)
modparam("dialog","profiles_with_value","outbound; inbound")
modparam("dialog", "dlg_match_mode", 1)
modparam("dialog", "default_timeout", 3600)
modparam("dialog", "options_ping_interval", 900)
modparam("dialog", "profiles_with_value", "caller ; domain")

```

**Comment by razvancrainea** [comment]
@volga629-1 are you also using the new `mi_script` module?

**Comment by volga629-1** [comment]
No

**Comment by razvancrainea** [comment]
could you please provide the full list of loaded modules, in the same order they have been loaded?

**Comment by volga629-1** [comment]
yes 

```
#auto_scaling_profile = PRF_TIMER
#     scale up to 10 on 70% for 4 cycles within 5   
#     scale down to 4 on 18% for 10 cycles
socket=udp:127.0.0.1:5060
#timer_workers=4
#listen=tls:
socket=tcp:
#listen=wss:
socket=smpp:
#listen=hep_udp:127.0.0.1:6090
#mhomed=1

udp_workers=10
tcp_accept_aliases=1
tcp_workers=10
#tcp_workers=32 use_auto_scaling_profile PRF_SIP 
tcp_max_msg_time=10
tcp_max_connections=8192
tcp_connection_lifetime=3600
tcp_keepalive=1
tcp_keepcount=3
tcp_keepidle=300
tcp_keepinterval=300
####### Modules Section ########

#set module path
mpath="/usr/lib64/opensips/modules/"

## TLS module 
#loadmodule "tls_mgm.so"
##modparam("tls_mgm", "db_url", "postgres://")
#modparam("tls_mgm", "server_domain", "default")
#modparam("tls_mgm","tls_method","[default]TLSv1")
#modparam("tls_mgm","verify_cert","[default]0")
#modparam("tls_mgm","require_cert","[default]0")
#modparam("tls_mgm","certificate","[default]/etc/opensips/tls/cert-prod.pem")
#modparam("tls_mgm","private_key","[default]/etc/opensips/tls/key-prod.pem")
#modparam("tls_mgm","ca_list","[default]/etc/opensips/tls/ca-prod.pem")
#modparam("tls_mgm","ca_dir","[default]/etc/opensips/tls")
#modparam("tls_mgm","ciphers_list", "[default]AES256-GCM-SHA384,AES256-SHA256,AES256-SHA,CAMELLIA256-SHA,AES128-SHA,SEED-SHA,CAMELLIA128-SHA,RC4-SHA,DES-CBC3-SHA")
#
## Client
#modparam("tls_mgm", "client_domain", "default_clnt")            
#modparam("tls_mgm", "certificate", "[default_clnt]/etc/opensips/tls/cert-prod.pem")          
#modparam("tls_mgm", "private_key", "[default_clnt]/etc/opensips/tls/key-prod.pem")       
#modparam("tls_mgm", "ca_list", "[default_clnt]/etc/opensips/tls/ca-prod.pem")      

# PgSQL module
loadmodule "db_postgres.so"

#loadmodule "tls_mgm.so"
#modparam("tls_mgm", "db_url", "postgres://")

#### SIGNALING module
loadmodule "signaling.so"

#### StateLess module
loadmodule "sl.so"

#### Transaction Module
loadmodule "tm.so"
modparam("tm", "fr_timeout", 5)
modparam("tm", "fr_inv_timeout", 30)
modparam("tm", "restart_fr_on_each_reply", 0)
modparam("tm", "onreply_avp_mode", 1)
modparam("tm", "timer_partitions", 4)

#### Record Route Module
loadmodule "rr.so"
/* do not append from tag to the RR (no need for this script) */
modparam("rr", "enable_double_rr", 1)
modparam("rr", "append_fromtag", 1)

#### Path module 
loadmodule "path.so"
modparam("path", "enable_double_path", 0)

#### MAX ForWarD module
loadmodule "maxfwd.so"

#### SIP MSG OPerationS module
loadmodule "sipmsgops.so"

# MI http
loadmodule "httpd.so"
loadmodule "mi_http.so"
loadmodule "event_stream.so"
#### Managment module
modparam("httpd", "ip", "127.0.0.1")
modparam("httpd", "port", 8092)
modparam("mi_http", "root", "mi")

#### FIFO Management Interface
loadmodule "mi_fifo.so"
modparam("mi_fifo", "fifo_name", "/run/opensips/opensips_fifo")
modparam("mi_fifo", "fifo_mode", 0666)

#### Group-based ACL Module
loadmodule "group.so"
modparam("group", "db_url", "postgres://")

#### SMS DID to user mapping
loadmodule "alias_db.so"
modparam("alias_db", "db_url", "postgres://")

#### UAC module to replace SMS DID From
loadmodule "uac_auth.so"
loadmodule "uac.so"
modparam("uac","restore_mode","manual")

#### USeR LOCation module
loadmodule "usrloc.so"
modparam("usrloc", "db_url", "postgres://")
modparam("usrloc", "working_mode_preset","single-instance-sql-write-through")
modparam("usrloc","nat_bflag","NATED_CALLEE")
modparam("usrloc", "desc_time_order", 1)
modparam("usrloc", "use_domain", 1)

#### REGISTRAR module
loadmodule "registrar.so"
modparam("registrar", "tcp_persistent_flag", "TCP_PERSIST_REGISTRATIONS")
modparam("registrar", "received_avp", "$avp(RECEIVED)")

#### AUTHentication modules
loadmodule "auth.so"
loadmodule "auth_db.so"
modparam("auth_db", "calculate_ha1", 1)
modparam("auth_db", "password_column", "password")
modparam("auth_db", "db_url","postgres://")
modparam("auth_db", "load_credentials", "")
modparam("auth_db", "use_domain", 1)

#### Domain Module
loadmodule "domain.so"
modparam("domain", "db_url","postgres://")
modparam("domain", "db_mode", 1)   # Use caching

#### Nat Helper
loadmodule "nathelper.so"
modparam("nathelper", "received_avp", "$avp(RECEIVED)")
modparam("nathelper", "natping_tcp", 1)
modparam("nathelper", "sipping_bflag", "SIPPING_START")
modparam("nathelper", "remove_on_timeout_bflag", "ZOMBIE_KILL_SWITCH")
modparam("nathelper", "max_pings_lost", 2)
modparam("nathelper", "natping_interval", 30)
modparam("nathelper", "natping_partitions", 6)
modparam("nathelper", "ping_threshold", 10)
modparam("nathelper", "sipping_ignore_rpl_codes", "408, 480, 404")
modparam("nathelper", "sipping_from", "sip:pinger@")
modparam("nathelper", "ping_nated_only", 1)

#### Nat Traversal
loadmodule "nat_traversal.so"
modparam("nat_traversal", "keepalive_state_file", "/run/opensips/keepalive_state")

#### ACCounting module
loadmodule "acc.so"
/* what special events should be accounted ? */
modparam("acc", "early_media", 0)
modparam("acc", "report_cancels", 0)
/* by default we do not adjust the direct of the sequential requests.
   if you enable this parameter, be sure the enable "append_fromtag"
   in "rr" module */
modparam("acc", "detect_direction", 0)

#### Cgrates
loadmodule "cgrates.so"
modparam("cgrates", "cgrates_engine", "127.0.0.1:2013")
modparam("cgrates", "bind_ip", "127.0.0.1")
modparam("cgrates", "max_async_connections", 20)

#### Freeswitch
loadmodule "freeswitch.so"

#### Dialog
loadmodule "dialog.so"
#modparam("dialog", "db_url", "postgres://")
modparam("dialog", "db_mode", 0)
modparam("dialog","profiles_with_value","outbound; inbound")
modparam("dialog", "dlg_match_mode", 1)
modparam("dialog", "default_timeout", 3600)
modparam("dialog", "options_ping_interval", 900)
modparam("dialog", "profiles_with_value", "caller ; domain")

#### Presence
#loadmodule "presence.so"
#loadmodule "presence_mwi.so"
#loadmodule "presence_xml.so"
#loadmodule "presence_dialoginfo.so"
#loadmodule "presence_callinfo.so"
#loadmodule "pua.so"
#loadmodule "pua_dialoginfo.so"
#loadmodule "xcap.so"
#modparam("presence|xcap|pua","db_url","postgres://")
##modparam("presence","server_address","sip:")
#modparam("presence", "clean_period",  30)
#modparam("presence", "mix_dialog_presence", 1)
#modparam("presence_xml", "force_active", 1)
#modparam("pua_dialoginfo", "presence_server", "sip:")
#modparam("pua_dialoginfo", "callee_spec_param", "$avp(blf_callee)")

#### Avpops
loadmodule "avpops.so"
modparam("avpops","db_url","postgres://")

#### Msilo
loadmodule "msilo.so"
modparam("msilo", "db_url", "postgres://")
modparam("msilo", "from_address", "sip:Offline@networklab.support")
modparam("msilo", "contact_hdr", "Contact: <sip:offline@networklab.support>\r\n")
modparam("msilo", "offline_message", "SMS Mailer:\nExt ~> $rU is offline.\nSMS accepted for offline delivery.\nSMS will be auto discarded in 24 hrs from ~> $time(%T).\n")
modparam("msilo", "content_type_hdr", "Content-Type: text/plain\r\n")
modparam("msilo", "expire_time", 86400)
modparam("msilo", "snd_time_avp", "$avp(snd_time)")
modparam("msilo", "outbound_proxy", "sip:127.0.0.1:5060")

#### Cache DB Redis
loadmodule "cachedb_redis.so"
modparam("cachedb_redis", "cachedb_url", "redis:wss-grp://127.0.0.1:6379/")
modparam("cachedb_redis", "cachedb_url", "redis:loc-sub://127.0.0.1:6379/")

#### SMPP protocol
loadmodule "proto_smpp.so"
modparam("proto_smpp", "db_url", "postgres://")
modparam("proto_smpp", "smpp_max_msg_chunks", 32)
modparam("proto_smpp", "smpp_send_timeout", 200)
modparam("proto_smpp", "smpp_port", 2776)
modparam("proto_smpp", "outbound_uri", "sip:127.0.0.1:5060")

### Lua script
loadmodule "lua.so"
modparam("lua", "luafilename", "/etc/opensips/scripts/convert.lua")

#### Rest client
loadmodule "rest_client.so"
 
#### Textops
loadmodule "textops.so"

#### UDP protocol
loadmodule "proto_udp.so"

#### TCP protocol
loadmodule "proto_tcp.so"

#### Tracing modules
#loadmodule "proto_hep.so"
#modparam("proto_hep", "hep_id", "[hid]:9060;transport=udp;version=3")
#loadmodule "tracer.so"
#modparam("tracer", "trace_id", "[tid]uri=hep:hid")
#
#### Tracing tls module
#loadmodule "proto_tls.so"
#modparam("proto_tls", "tls_max_msg_chunks", 16)

#### Websocket
#loadmodule "proto_wss.so"
#loadmodule "proto_ws.so"
#modparam("proto_wss", "wss_max_msg_chunks", 16)

####### Routing Logic ########

# main request routing logic
```

**Comment by razvancrainea** [comment]
Is this only happening with postgres 12.6, as that's what the initial report suggests. Have you tried with different DBs as well?

Best regards,
Răzvan

**Comment by volga629-1** [comment]
I didn't tried different db variant, but PgSQL is affected I tested 12 and 11 versions.

**Comment by bogdan-iancu** [comment]
@volga629-1 , please provide here (as attachement) the full opensips logs in debug level (set log_level to 4)

**Comment by volga629-1** [comment]
Here are start up log
[opensips-startup-9Jul2021.txt](https://github.com/OpenSIPS/opensips/files/6791529/opensips-startup-9Jul2021.txt)

**Comment by bogdan-iancu** [comment]
Thanks @volga629-1 , could you re-do this but first compiling in this small patch ?
```
diff --git a/modules/dialog/dlg_db_handler.c b/modules/dialog/dlg_db_handler.c
index f9ba84fbf..b27e963cf 100644
--- a/modules/dialog/dlg_db_handler.c
+++ b/modules/dialog/dlg_db_handler.c
@@ -160,6 +160,7 @@ int dlg_connect_db(const str *db_url)
        }
        if ((dialog_db_handle = dialog_dbf.init(db_url)) == 0)
                return -1;
+       LM_DBG("----- DB con created at %p\n",dialog_db_handle);
        return 0;
 }
 
@@ -230,6 +231,7 @@ int init_dlg_db(const str *db_url, int dlg_hash_size , int db_update_period)
 
        dialog_dbf.close(dialog_db_handle);
        dialog_db_handle = 0;
+       LM_DBG("----- DB con reset to %p\n",dialog_db_handle);
 
        return 0;
 }
@@ -242,6 +244,7 @@ void destroy_dlg_db(void)
        if (dialog_db_handle) {
                dialog_dbf.close(dialog_db_handle);
                dialog_db_handle = 0;
+               LM_DBG("----- DB con reset to %p\n",dialog_db_handle);
        }
 }
```

**Comment by volga629-1** [comment]
Yes let me do it

**Comment by volga629-1** [comment]
Patch failed 
I will try add manually 

```
Patch #995 (opensips-3.2-beta-dialog.patch):
+ /usr/bin/patch --no-backup-if-mismatch -p1 -b --suffix .dialog --fuzz=0
patching file modules/dialog/dlg_db_handler.c
Hunk #1 FAILED at 160.
Hunk #2 FAILED at 230.
Hunk #3 FAILED at 242.
3 out of 3 hunks FAILED -- saving rejects to file modules/dialog/dlg_db_handler.c.rej
error: Bad exit status from /var/tmp/rpm-tmp.dp6eKy (%prep)

```

**Comment by volga629-1** [comment]
Here are startup with patch

[opensips-startup-patched-9Jul2021.txt](https://github.com/OpenSIPS/opensips/files/6792111/opensips-startup-patched-9Jul2021.txt)

**Comment by bogdan-iancu** [comment]
@volga629-1 , as pointed out by @liviuchircu , this is already fixed in 57df749c7d2b3cc47d82b05a4bd4492e76e758fd , so please use the latest sources or the nightly builds (for packages)/
